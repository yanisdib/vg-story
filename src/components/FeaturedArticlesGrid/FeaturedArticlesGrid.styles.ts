import styled from 'styled-components'


export const Section = styled.section`
    --ui-grid-mg-y: 100px;
    --card-image-mg-y: 20px;

    display: flex;
    flex-direction: row;
    height: auto;
    margin: var(--ui-grid-mg-y) 0;
`;

export const Wrapper = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 100%;
`;

export const Content = styled.div`
    display: flex;
    flex-direction: row;
    width: 100%;
    height: 100%;
`;

export const Grid = styled.div`
    display: grid;
    grid-template: auto;
    row-gap: var(--ui-grid-mg-y);
    column-gap: var(--ui-grid-mg-y);
    justify-items: center;
    width: 100%;
    height: 100%;

    h3, h5 {
        margin: 8px 0;
    }

    p {
        margin: 15px 0;
        line-height: 28px;
        font-size: 1.2em;
        font-weight: 200;
        color: var(--ui-shark-color);
    }
`;

export const Card = styled.div`
    display: flex;
    flex-direction: column;
    max-width: 100%;
    height: fit-content;

    img {
        width: 100%;
        height: 100%;
        object-fit: contain;
        object-position: center;
        margin-bottom: var(--card-image-mg-y);
    }
`;

export const FullCard = styled.div`
    display: flex;
    flex-direction: row;
    grid-column: span 2;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: fit-content;

    div {
        width: 100%;
        height: 100%;
        flex-basis: 54.5%;

        &:first-child {
            flex-basis: 37.5%;
            margin-right: 8%;
        }
    }

    img {
        width: 100%;
        height: 100%;
        flex-basis: 54.5%;
        object-fit: cover;
        object-position: center;
    }
`;

export const Column = styled.div``;

export const ViewMore = styled.div`
    display: flex;
    flex-direction: row;
    margin-top: var(--ui-grid-mg-y);

    h3 {
        font-family: var(--ui-subtitle-font);
        font-weight: 500;
    }
`;