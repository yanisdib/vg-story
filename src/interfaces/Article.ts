import {
    Comment,
    ImageAttributes,
    Topic
} from '.';


/**
 * Article is an interface which describes properties
 * an article should be created with
 */
export interface Article {
    author: string;
    backgroundHex?: string;
    body?: string;
    comments?: Comment[];
    createdAt: number;
    editedAt?: number;
    frontImage: ImageAttributes;
    imageGallery?: {
        id: string,
        images: ImageAttributes[]
    }[];
    id: string;
    isFeatured?: boolean;
    permalink: string;
    title: string;
    topics: Topic[];
}