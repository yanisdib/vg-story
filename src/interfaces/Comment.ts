/**
 * Comment is an interface which describes properties
 * an article's comment should be created with
 */
export interface Comment {
    id: string;
    body: string;
    createdAt: number;
}