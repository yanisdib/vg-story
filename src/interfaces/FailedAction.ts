import { Action } from '@reduxjs/toolkit';


/**
 * FailedAction is an interface which describes
 * properties returned by the payload of
 * an action that failed
 */
export interface FailedAction extends Action {
    error: Error;
}