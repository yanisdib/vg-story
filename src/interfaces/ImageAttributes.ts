/**
 * Attributes an image object should be declared with
 */
export interface ImageAttributes {
    alt?: string;
    src: string;
    title?: string;
}