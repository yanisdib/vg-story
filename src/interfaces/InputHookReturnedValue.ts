export type InputHookReturnedValue<T> = [
    T,
    (event: React.ChangeEvent<HTMLInputElement>) => void
]
