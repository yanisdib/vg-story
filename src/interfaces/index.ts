export type { Article } from './Article';
export type { Comment } from './Comment';
export type { ImageAttributes } from './ImageAttributes';
export type { Topic } from './Topic';

export type { FailedAction } from './FailedAction';
export type { StoreState } from './StoreState';
