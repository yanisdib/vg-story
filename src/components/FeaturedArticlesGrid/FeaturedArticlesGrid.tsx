import { Link } from 'react-router-dom';

import { Article } from '../../interfaces';

import Container from '../Container/Container';

import {
    Card,
    Column,
    Content,
    FullCard,
    Grid,
    Section,
    ViewMore,
    Wrapper
} from './FeaturedArticlesGrid.styles';


interface FeaturedArticlesGridProps {
    featuredArticles: Article[];
}


function FeaturedArticlesGrid(
    { featuredArticles }: FeaturedArticlesGridProps
): JSX.Element {
    return (
        <Section>
            <Wrapper>
                <Container>
                    <Content>
                        <Grid>
                            {featuredArticles
                                .flatMap((article, index) => {

                                    const {
                                        body,
                                        frontImage,
                                        id,
                                        permalink,
                                        title,
                                        topics
                                    } = article;

                                    return (index < 2) ? (
                                        <Card key={id}>
                                            <Link to={`article/${permalink}`}>
                                                <img
                                                    alt={frontImage?.alt ?? title}
                                                    src={frontImage.src}
                                                    title={frontImage?.title ?? title}
                                                />
                                            </Link>
                                            <h5>
                                                {topics
                                                    .map(topic => topic.name)
                                                    .join(' — ')}
                                            </h5>
                                            <Link to={`article/${permalink}`}>
                                                <h3>{title}</h3>
                                            </Link>
                                            <p>{body}</p>
                                        </Card>
                                    ) : (
                                        <FullCard key={id}>
                                            <Column>
                                                <h5>
                                                    {topics
                                                        .map(topic => topic.name)
                                                        .join(' — ')}
                                                </h5>
                                                <Link to={`article/${permalink}`}>
                                                    <h3>{title}</h3>
                                                </Link>
                                                <p>{body}</p>
                                            </Column>
                                            <Column>
                                                <Link to={`article/${permalink}`}>
                                                    <img
                                                        alt={frontImage?.alt ?? title}
                                                        src={frontImage.src}
                                                        title={frontImage?.title ?? title}
                                                    />
                                                </Link>
                                            </Column>
                                        </FullCard>
                                    );
                                })}
                        </Grid>
                    </Content>
                </Container>
                <ViewMore>
                    <Link to="/browse/articles">
                        <h3>View all stories -&gt;</h3>
                    </Link>
                </ViewMore>
            </Wrapper>
        </Section>
    );
}


export default FeaturedArticlesGrid;