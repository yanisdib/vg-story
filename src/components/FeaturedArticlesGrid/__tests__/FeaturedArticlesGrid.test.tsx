import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { Article } from '../../../interfaces';

import FeaturedArticlesGrid from '../FeaturedArticlesGrid';


describe('FeaturedArticlesGrid component', () => {

    const featuredArticles: Article[] = [
        {
            author: 'Yanis',
            backgroundHex: '#FF5B5B',
            frontImage: { src: 'hello.jpg' },
            createdAt: 1000,
            id: '12de',
            permalink: 'this-is-a-test',
            title: 'This is a test',
            topics: [
                {
                    createdAt: 0,
                    id: '1ac',
                    name: 'Test',
                    permalink: 'test'
                }
            ]
        },
        {
            author: 'Jean',
            backgroundHex: '#8C5BFF',
            frontImage: { src: 'hello.jpg' },
            createdAt: 1000,
            id: '12fe',
            permalink: 'this-is-a-second-test',
            title: 'This is a second test',
            topics: [
                {
                    createdAt: 0,
                    id: '1ac',
                    name: 'Test',
                    permalink: 'test'
                }
            ]
        },
        {
            author: 'Pierre',
            backgroundHex: '#2848FF',
            frontImage: { src: 'hello.jpg' },
            createdAt: 1000,
            id: '1he',
            permalink: 'this-is-a-third-test',
            title: 'This is a third test',
            topics: [
                {
                    createdAt: 0,
                    id: '1ac',
                    name: 'Test',
                    permalink: 'test'
                }
            ]
        }
    ];

    it('should render without crashing', () => {

        const { container } = render(
            <FeaturedArticlesGrid featuredArticles={featuredArticles} />,
            { wrapper: MemoryRouter }
        );

        expect(container).toMatchSnapshot();
    });

    it('should render two Card components with the two latest featured articles', () => {

        const { getByTestId } = render(
            <FeaturedArticlesGrid featuredArticles={featuredArticles} />,
            { wrapper: MemoryRouter }
        );

        featuredArticles.forEach((article, index) => {
            (index < featuredArticles.length - 2) && (
                expect(getByTestId(`card-test-${index}`))
                    .toBeInTheDocument()
            );
        });
    });

    it('should render a FullCard component if there is more than 3 featured articles', () => {

        const { getByTestId } = render(
            <FeaturedArticlesGrid featuredArticles={featuredArticles} />,
            { wrapper: MemoryRouter }
        );

        expect(getByTestId('fullcard-test')).toBeInTheDocument();
        expect(getByTestId('fullcard-test')).toHaveStyle('grid-column: span 2');
    });
});