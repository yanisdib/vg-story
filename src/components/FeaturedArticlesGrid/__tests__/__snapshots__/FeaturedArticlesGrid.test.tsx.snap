// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeaturedArticlesGrid component should render without crashing 1`] = `
<div>
  <section
    class="sc-kgflAQ HsGpL"
  >
    <div
      class="sc-fLlhyt eUxozV"
    >
      <div
        class="sc-bczRLJ bJuwle"
      >
        <div
          class="sc-bBrHrO eiyQrd"
        >
          <div
            class="sc-ivTmOn cPpDWF"
          >
            <div
              class="sc-cxabCf jlMXVw"
              data-testid="card-test-0"
            >
              <a
                href="/article/this-is-a-test"
              >
                <img
                  alt="This is a test"
                  src="hello.jpg"
                  title="This is a test"
                />
              </a>
              <h5>
                Test
              </h5>
              <a
                href="/article/this-is-a-test"
              >
                <h3>
                  This is a test
                </h3>
              </a>
              <p />
            </div>
            <div
              class="sc-cxabCf jlMXVw"
              data-testid="card-test-1"
            >
              <a
                href="/article/this-is-a-second-test"
              >
                <img
                  alt="This is a second test"
                  src="hello.jpg"
                  title="This is a second test"
                />
              </a>
              <h5>
                Test
              </h5>
              <a
                href="/article/this-is-a-second-test"
              >
                <h3>
                  This is a second test
                </h3>
              </a>
              <p />
            </div>
            <div
              class="sc-llJcti gpiUky"
              data-testid="fullcard-test"
            >
              <div
                class="sc-iIPllB"
              >
                <h5>
                  Test
                </h5>
                <a
                  href="/article/this-is-a-third-test"
                >
                  <h3>
                    This is a third test
                  </h3>
                </a>
                <p />
              </div>
              <div
                class="sc-iIPllB"
              >
                <a
                  href="/article/this-is-a-third-test"
                >
                  <img
                    alt="This is a third test"
                    src="hello.jpg"
                    title="This is a third test"
                  />
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-gicCDI hmRkji"
      >
        <a
          href="/browse/articles"
        >
          <h3>
            View all stories -&gt;
          </h3>
        </a>
      </div>
    </div>
  </section>
</div>
`;
